### ISSUE TITLE
MMSeg model in Docker `mmdeploy.apis.pytorch2onnx.torch2onnx` with Call id: 0 failed. exitt

### ISSUE THREAD

**Issue by davodogster** [open-mmlab/mmdeploy#1369]
### Checklist

- [X] I have searched related issues but cannot get the expected help.
- [X] 2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/faq.md) but cannot get the expected help.
- [X] 3. The bug has not been fixed in the latest version.

### Describe the bug

MMSeg model in Docker `mmdeploy.apis.pytorch2onnx.torch2onnx` with Call id: 0 failed. exit.

### Reproduction

Hi guys, @lvhan028 

Do you know why this is failing? I trained the model in MMSegmentation Docker container:
![image](https://user-images.githubusercontent.com/46079516/202036904-3e639cab-375c-4d4a-9f10-d727343f84ab.png)

Now I'm running the MMDeploy CPU Docker container and it gives an error.
When I enter the container I run pip install mmsegmentation
![image](https://user-images.githubusercontent.com/46079516/202036624-d29b66cb-2236-42b4-827f-3883ddaca4dc.png)

### Environment

```Shell
2022-11-15 22:32:49,270 - mmdeploy - INFO -

2022-11-15 22:32:49,270 - mmdeploy - INFO - **********Environmental information**********
fatal: not a git repository (or any of the parent directories): .git
2022-11-15 22:32:49,477 - mmdeploy - INFO - sys.platform: linux
2022-11-15 22:32:49,477 - mmdeploy - INFO - Python: 3.7.13 (default, Mar 29 2022, 02:18:16) [GCC 7.5.0]
2022-11-15 22:32:49,477 - mmdeploy - INFO - CUDA available: False
2022-11-15 22:32:49,477 - mmdeploy - INFO - GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
2022-11-15 22:32:49,477 - mmdeploy - INFO - PyTorch: 1.10.0+cpu
2022-11-15 22:32:49,477 - mmdeploy - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX512
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=0, USE_CUDNN=OFF, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=ON, USE_OPENMP=ON,

2022-11-15 22:32:49,478 - mmdeploy - INFO - TorchVision: 0.11.0+cpu
2022-11-15 22:32:49,478 - mmdeploy - INFO - OpenCV: 4.5.3-openvino
2022-11-15 22:32:49,478 - mmdeploy - INFO - MMCV: 1.5.3
2022-11-15 22:32:49,478 - mmdeploy - INFO - MMCV Compiler: GCC 7.3
2022-11-15 22:32:49,478 - mmdeploy - INFO - MMCV CUDA Compiler: not available
2022-11-15 22:32:49,478 - mmdeploy - INFO - MMDeploy: 0.10.0+
2022-11-15 22:32:49,478 - mmdeploy - INFO -

2022-11-15 22:32:49,478 - mmdeploy - INFO - **********Backend information**********
2022-11-15 22:32:49,929 - mmdeploy - INFO - onnxruntime: 1.8.1  ops_is_avaliable : True
2022-11-15 22:32:49,931 - mmdeploy - INFO - tensorrt: None      ops_is_avaliable : False
2022-11-15 22:32:49,963 - mmdeploy - INFO - ncnn: 1.0.20221104  ops_is_avaliable : True
2022-11-15 22:32:49,965 - mmdeploy - INFO - pplnn_is_avaliable: False
2022-11-15 22:32:49,991 - mmdeploy - INFO - openvino_is_avaliable: True
2022-11-15 22:32:50,018 - mmdeploy - INFO - snpe_is_available: False
2022-11-15 22:32:50,020 - mmdeploy - INFO - ascend_is_available: False
2022-11-15 22:32:50,022 - mmdeploy - INFO - coreml_is_available: False
2022-11-15 22:32:50,022 - mmdeploy - INFO -

2022-11-15 22:32:50,022 - mmdeploy - INFO - **********Codebase information**********
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmdet:      None
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmseg:      0.29.1
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmcls:      0.24.1
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmocr:      None
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmedit:     None
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmdet3d:    None
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmpose:     None
2022-11-15 22:32:50,024 - mmdeploy - INFO - mmrotate:   None
```

**Comment by lvhan028** [comment]
@RunningLeon could you please investigate this issue?

**Comment by RunningLeon** [comment]
@davodogster Hi
Q1: dynamic onnxruntime failed
Something wrong with your checkpoint, pls. check it.
Q2:  FCN_Unet model failed.
Something wrong with your  custom deploy config, you need to have base config like https://github.com/open-mmlab/mmdeploy/blob/99040d5655aff9b39f331db925dce0c065a8176a/configs/mmseg/segmentation_onnxruntime_static-1024x2048.py#L1 pls post your full deploy config here. Besides, suggest you directly change deploy config provided by mmdeploy.

**Comment by davodogster** [comment]
@RunningLeon

Q1: I think the checkpoint is fine as I was able to load it and use it for inference previosuly but I will double check it.

Q2: FCN_Unet only failed because of "type field" but I already specified a type field using this custom config: 

`onnx_config = dict( type='onnx', dynamic_axes={ 'input': { 0: 'batch', 2: 'height', 3: 'width' }, 'output': { 0: 'batch' } }, export_params=True, keep_initializers_as_inputs=False, opset_version=11, save_file='end2end.onnx', input_names=['input'], output_names=['output'], input_shape=[992, 992])`

**Comment by RunningLeon** [comment]
> Thanks @RunningLeon
> 
> Q1: I think the checkpoint is fine as I was able to load it and use it for inference previosuly but I will double check it.
> 
> Q2: FCN_Unet only failed because of "type field" but I already specified a type field using this custom config:
> 
> `onnx_config = dict( type='onnx', dynamic_axes={ 'input': { 0: 'batch', 2: 'height', 3: 'width' }, 'output': { 0: 'batch' } }, export_params=True, keep_initializers_as_inputs=False, opset_version=11, save_file='end2end.onnx', input_names=['input'], output_names=['output'], input_shape=[992, 992])`

Hi, 
Q2: if you posted is the full content of the custom deploy config, then it's not right. You have to set the codbease_config and backend_config keys in the config.py. As mentioned before, pls. change according to the example config. https://github.com/open-mmlab/mmdeploy/blob/99040d5655aff9b39f331db925dce0c065a8176a/configs/mmseg/segmentation_onnxruntime_static-1024x2048.py#L1

**Comment by davodogster** [comment]
@RunningLeon  Thanks I will try that now. And if I trained on 992x992 do I need to specify input_shape=[992, 992] and also change the file name to end with -992x992.py? Will it work? 

I'm thinking of training a Transformer network next and then converting it to ONNX then TRT.

**Comment by davodogster** [comment]
@RunningLeon Thanks the model conversion worked after I modified segmentation_onnxruntime_static-1024x2048.py with my 512, 1024 but since I am using a custom dataset (named External Dataset) it throws an error for 1. visualization and 2. Deployment. this is frustrating. Do you know how I can fix it so that I can test the new onnx model?

![image](https://user-images.githubusercontent.com/46079516/202594296-9be6ffce-4d9f-4ead-827e-42504612e399.png)

![image](https://user-images.githubusercontent.com/46079516/202594260-b006a339-cfba-4b99-a380-42c6d03b8bd8.png)

Thanks, Sam

**Comment by Masque555** [comment]
> @RunningLeon could you please investigate this issue?
Hi，Is the onnx model allowed to input channel dynamic mode? 
For example, if the onnx input dimension is 1 * 3 * 1000 * 800, I can change the input dimension to 1 * 1 * 1000 * 800?
thank you!

**Comment by RunningLeon** [comment]
> 

Hi， LD_LIBRARY_PATH is not right with mmdeploy `::/root/workspace/mmdeploy/lib`. Besides, `$ONNXRUNTIME_DIR/lib` is missing. Could refer to [export LD_LIBRARY_PATH=$ONNXRUNTIME_DIR/lib:$LD_LIBRARY_PATH](https://mmdeploy.readthedocs.io/en/latest/05-supported-backends/onnxruntime.html#prerequisite)

**Comment by Masque555** [comment]
> > > @RunningLeon could you please investigate this issue?
> > > Hi，Is the onnx model allowed to input channel dynamic mode?
> > > For example, if the onnx input dimension is 1 * 3 * 1000 * 800, I can change the input dimension to 1 * 1 * 1000 * 800?
> > > thank you!
> 
> No. channel is not dynamic .
Thanks for your reply. What should I do if I want the input channel to be 1.

**Comment by RunningLeon** [comment]
@davodogster Hi, change to `input_shape = [2048, 1024]` if you are using a static deploy config and want to evaluate it with image 1024x2048.

**Comment by davodogster** [comment]
> @davodogster Hi, change to `input_shape = [2048, 1024]` if you are using a static deploy config and want to evaluate it with image 1024x2048.

Hi @RunningLeon I just want to test the model on whatever it was designed to inference on (this is why I posted my model config above :) MMSeg training and testing shapes are very confusing. Are the the same?

I tried [1024, 2028] and [2048, 1024] and I am still getting errors. Do you think the MMDeploy Documentation could have some more details? 

![image](https://user-images.githubusercontent.com/46079516/203154230-f435bc04-b645-4391-a6bf-547804d8da3e.png)

Thanks, Sam

**Comment by RunningLeon** [comment]
@davodogster Hi, if you want to deploy and test with image shape of 1 x 3 x 1024 x 2048 in B x C x H x W, pls. try
1. change `img_scale` of `val` pipeline to `img_scale=[2048, 1024] # width, height` and `Resize` of `transform` to `dict(type='Resize', keep_ratio=False),`
2. Convert to onnx use deploy config of `configs/mmseg/segmentation_onnxruntime_static-1024x2048.py`, in which the `onnx_config` is set to `onnx_config = dict(input_shape=[2048, 1024]) # width, height`
3. use `tools/test.py` to evaluate onnx model with onnxruntime backend.

**Comment by davodogster** [comment]
@RunningLeon Thanks I'll try that. Was my model trained to do inference on images size 2048, 1024? What is the difference between train crop size and test img_scale?

Also, I am running the GPU container for TRT. Do you how to convert the FCN Unet ONNX model to TRT? It is throwing an error about the node Argmax_97
`trtexec --onnx=end2enddynamic.onnx --shapes=input:1x3x768x768 --saveEngine=FM_fcn_unet_engine.trt # shape??
`
![image](https://user-images.githubusercontent.com/46079516/203218701-6c0b0c0f-c14f-41ff-9016-38def29b7f96.png)

**Comment by RunningLeon** [comment]
@davodogster 
1. image would be cropped with `crop_size` in training mode, while would use `img_scale` for evaluation.
2. upgrade tensorrt >= 8.2

**Comment by davodogster** [comment]
Hi @RunningLeon I was using TRT 8.2.4 when it gave the error:
Its the MMDeploy GPU Docker container.
![image](https://user-images.githubusercontent.com/46079516/203424546-975a851d-dc90-4d69-a68e-46c7b2d2ee6c.png)

**Comment by RunningLeon** [comment]
Hi, try enlarge the `max_workspace_size` in deploy config, note that `1>>30` equals to `1GB`.
https://github.com/open-mmlab/mmdeploy/blob/de96f51231d9ea900dbc42fb63cbb63b7d60cc71/configs/mmseg/segmentation_tensorrt_static-1024x2048.py#L5

**Comment by davodogster** [comment]
@RunningLeon thanks that worked! I set it in the trtexec flags

`trtexec --onnx=fcn-resnet101.onnx --fp16 --workspace=64 --minShapes=input:1x3x256x256 --optShapes=input:1x3x1026x1282 --maxShapes=input:1x3x1440x2560 --buildOnly --saveEngine=fcn-resnet101.engine
`

DO you know why MMDeploy TRT version will not install below? I think I have done everything correctly and downloaded the .tar.gz from NVIDIA dev website.
![image](https://user-images.githubusercontent.com/46079516/203462664-d5696bc0-fa18-482d-a95e-7921c4a67828.png)

**Comment by davodogster** [comment]
@RunningLeon sorry, I Solved it by using the CP 39.whl file instead of 38 :)

pip install /home/scion.local/davidsos/Documents/mmdeploy_trt/TensorRT-8.2.3.0/python/tensorrt-8.2.3.0-cp39-none-linux_x86_64.whl

Maybe there could ne a note about it in the instructions? Cheers

**Comment by davodogster** [comment]
@RunningLeon Thanks I'll check that.

What is the difference between:
1. convert .pth --> ONNX --> TRT Vs
2. convert .pth to TRT using:

![image](https://user-images.githubusercontent.com/46079516/203677114-8fbeed44-69b7-4a8c-b64d-0a2b01ede4c4.png)

3. When converting to TRT using either method, is it essential to have a calibration dataset?
"a custom dataset config can be given to calibrate the model" config "can" be given or "must" be given?

Cheers, Sam

**Comment by RunningLeon** [comment]
1. for trt , we only provide pth->onnx->trt
3. calibration dataset is for trt-int8. if you want to enable int8, you have to set a proper calibration dataset.

**Comment by davodogster** [comment]
@RunningLeon ok thanks, 

This says MMDet to TRT. Does it automatically convert to ONNX as an intermediate step?
![image](https://user-images.githubusercontent.com/46079516/203686441-eeac5078-7a41-4dc7-914e-40ee3a3491c0.png)

**Comment by davodogster** [comment]
The MMDeploy GPU Docker container comes with TRT 8.2.4, but the mmdeploy inference instructions are for mmdep 8.2.3.

Is there a GPU container with TRT 8.2.3? Or do I need to redo this installation with 8.2.4?
![image](https://user-images.githubusercontent.com/46079516/203693557-15e1c4d2-6511-4eae-8e27-d27a5fe77ccd.png)

(Converting in docker, inferencing in pure python)

So the model deployment/inference fails.
![image](https://user-images.githubusercontent.com/46079516/203693676-8e948fed-05e2-4ff8-afe5-498369344323.png)


Thanks, Sam

**Comment by RunningLeon** [comment]
Hi, you do not need to install mmdeploy if you use docker. When you build docker image, the mmdeploy would be built automatically as show here.
https://github.com/open-mmlab/mmdeploy/blob/301035a06f5d298bd117c277f9fcb9b9dd292e5e/docker/GPU/Dockerfile#L77
